### ISSUE TITLE
Implement two-finger pan and pinch zoom

### ISSUE THREAD

**Issue by tonglianheng** [otfried/ipe#91]
Hi,

I wish to report the following issue I have encountered with using Ipe.

System versions:

Ipe version:   7.2.5 prebuilt Mac OS X binary downloaded from ipe.otfried
Mac OS X:   El Capitan 10.11.6

Issue:

When drawing with the multi-touch track pad, touching the trackpad with two fingers will always jump the canvas to centre at the bottom left corner (0,0).  This means that if the two finger click is setup (as standard) on Mac OS X as mouse-right-click, then the cursor will always jump to some point near the bottom left corner of the canvas.  This makes the two finger click for finishing a line unusable as the object will always terminate somewhere near (0,0).  And whenever I select the property of an object, the object will be jumped out of the view.

This appears to be an issue related with zooming, because after I have set (in Mac OS X preference) mouse-right-click to be the click at the bottom right corner of the trackpad, the right-click functions all work perfectly.  However, zooming by two finger "wheel" action still sends the canvas to centre at bottom left corner.  I guess the ideal behaviour should be that the zooming action should be centred around where the mouse pointer currently is or at the centre of the current view screen, but somehow it ended up at zooming around point (0,0) of the canvas.

**Comment by tonglianheng** [comment]
Okay, looking in action.lua, the pan position for zooming looks like to be taken as the previous pan position which I guess is initialised to (0,0), and after I pressed "x" to set pan, two fingers touch will then centre around that point.

So I guess I will just have remember to set pan using "x" at the beginning.

**Comment by tonglianheng** [comment]
After I have opened a new session, pressing "x" for the first time, and then two-finger touch the touch pad, the canvas jumps to (0,0) again.  And I can no longer reproduce the behaviour in my previous comment.  Two fingers touch seems to always reset to (0,0), rendering it almost useless.

**Comment by tonglianheng** [comment]
Okay, I found that press holding "x" while changing the zoom with two fingers repeatedly while trying to keep the cursor at one position of the canvas for a while, and you will notice zooming will no longer jump to (0,0), and after that the correct behaviour ensues.

EDIT:

Actually after a closer look press holding "x" while changing the zoom with two fingers just changes the "origin" of where a touch by two fingers resets to the next time---the canvas will still jump, but will jump to the new point.

**Comment by tianyikillua** [comment]
I can confirm this error and it is indeed very annoying. Could use double simple click to end the line drawing, but I'm used to use two fingers to do that...

**Comment by otfried** [comment]
Did this work correctly in 7.2.2? 

I started working on handling the zoom gesture on OSX in 7.2.3, but then I lost access to my trackpad and could not finish this.   I thought I had reverted to a working state, but apparently something is still wrong.

I will not have access to a trackpad until 2017, so I'll need help to patch this.

**Comment by otfried** [comment]
If you can compile Ipe yourself, try changing Canvas::scrollWheel (in _ipecanvas_cocoa.cpp_) like this
(that is, comment out the part dealing with panning):

```
void Canvas::scrollWheel(NSEvent *event)
{
  if (!iObserver)
    return;
  /*
  Vector v(event.scrollingDeltaX, -event.scrollingDeltaY);
  // NSLog(@"scroll: %lu %g %g", event.phase, v.x, v.y); 
  v = (1.0/zoom()) * v;
  if (event.phase == NSEventPhaseBegan || event.phase == NSEventPhaseNone)
    iGesturePan = pan();
  setPan(iGesturePan - v);
  update();
  */
  double delta = [event scrollingDeltaY];
  bool precise = [event hasPreciseScrollingDeltas];
  iObserver->canvasObserverWheelMoved(delta, precise);
}
```

**Comment by tonglianheng** [comment]
Hi Otfried,

Thank you very much for your reply.

I will try to compile the source on my mac and see if this fixes the issue.

Best regards,

Lianheng

**Comment by tonglianheng** [comment]
Hi Otfried,

I don't have a binary for 7.2.2, if you have one I can test it too.

**Comment by otfried** [comment]
https://bintray.com/otfried/generic/ipe/7.2.2#files/ipe/7.2

**Comment by tonglianheng** [comment]
Hi Otfried,

I have visited the link before, but there are no files to download?

Lianheng

**Comment by otfried** [comment]
Click on "Files".

**Comment by tonglianheng** [comment]
Great, got it now.

## 

I can confirm that 7.2.2 works.  The two finger touch zoom still works, but pan no longer jumps.  I like this behaviour for the scrollwheel zooming.

So if I understood correctly what you have said, after 7.2.2 you tried to implement the "pinch zoom"?

**Comment by otfried** [comment]
Yes. My understanding is that OSX doesn't use the scroll wheel zoom anywhere, so i wanted to implement pinch zoom and two finger pan.

**Comment by tonglianheng** [comment]
Google maps still uses the scroll wheel zoom.

One potential issue I see with two finger pan is that it can clash with two-finger tab/click, which on default is set to be equivalent to right-click and I think is what most people got used to.

So I suggest you ignore panning all together, and just zoom with whatever pan it is currently set.  And this ought to work.

I quite like the idea of panning just with "x".

**Comment by tonglianheng** [comment]
So should I go ahead and compile the 7.2.5, or is it not necessary and you know what to fix?

**Comment by otfried** [comment]
I don't think you need to try compiling - I'm pretty sure I know how to fix the problem now.

In 7.2.6 (in September) I will revert to how it worked in 7.2.2.    Later (in 2017) we can discuss if this is the best implementation for OSX users.

Isn't there plenty of software where two-finger gestures are used both for right-click and for panning?  I thought it distinguishes the gestures simply by checking if the fingers were moved while in contact with the trackpad.

**Comment by tonglianheng** [comment]
Ah, okay I understand what you mean now.  Yes, it does distinguish by checking if fingers moved after contact.  This is how the scrolling works anyway.

I actually just came across Ipe a few days before, after searching for a xfig replacement on Mac OS X.  And after a couple of days using it, I really like it!  Are you the only developer for the project?

If you want, I am willing to learn and help with developing some of its features.  I am currently a developer for the molecular structure software CP2K, but I am moving to a new job in September and this will will give me may be more time to play around new codes.

**Comment by m7thon** [comment]
I wanted to chime in. I only ever use a Mac with a touchpad, so here is my opinion on the most natural user interface:
- 2-finger clicking is almost always set up to be a right-click
- moving with 2 fingers means pan in almost all applications (yes, Google Earth is different, very annoying), not zoom
- zoom is typically done by the pinch gesture

I also think that this makes a lot of sense for ipe. At least in my case, I like to zoom and pan around a lot, so having both available via the standard gestures would help. I currently run ipe 7.2.2 which uses scroll-wheel zoom, and this is indeed very annoying on a touch pad. Also, at least in ipe 7.2.2, the speed of zoom is very difficult to control when using the 2-finger-move gesture (wheel-zoom).

**Comment by otfried** [comment]
Ultimately I want to implement it the standard way: two-finger tap is right-click, two-finger drag is pan, pinch gesture to zoom.

However, I do not have access to a trackpad right now (probably again in 2017), so for the next release we'll simply revert to how it worked in 7.2.2 (fixing the jumping).

**Comment by otfried** [comment]
@m7thon By the way, it seems macports is still stuck at 7.2.2.  Is there any reason for that?  Bug #69 (Apple's autocorrect for quotes) makes Ipe 7.2.2 kind of unusable for anyone doing real math in Ipe on OSX.

**Comment by m7thon** [comment]
@otfried Well, MacPorts likes to have ports that are backwards compatible on older versions of OSX also. That makes it a little tricky to update ipe, since ipe now uses Objective-C features that are only available I think for OSX >= 10.10.

I could not reproduce bug #69.

I think now is a bad time for updating because of the transition to Mac OS Sierra. Also, I think one of the most important usability improvements is this current issue (scrolling and panning), so I thought a good time to update would be the next ipe version. 

Finally, since you now provide Mac binary versions, I am not sure how relevant it is to keep maintaining a MacPorts version.

I have not seen any complaints regarding the MacPorts version (either on ipe-discuss, here, or MacPorts trac), so either people are not using the MacPorts version, or there seems to be little need for update at this time.

**Comment by otfried** [comment]
@m7thon Well, I just had a bug report (by email) from a macports user, that's why I was asking. It's a bit annoying to have to respond about bugs I've already fixed.    I'll tell him to switch to my package then. 

Bug #69 may be locale-dependent (as quotes are language dependent).  I had no difficulty reproducing it on OSX 10.11 set to English.

I have not introduced now Objective-C dependencies after 7.2.2, and applied most of your patch for #60, so that shouldn't be a problem. 

The current issue is "fixed" in 7.2.6 in the sense that it doesn't "jump" anymore.   In the long run, I plan to support pinch zoom and two-finger pan, but I cannot give any time prediction for that (I'll have a trackpad again in 2017, but then I'll also be teaching again).

Personally I don't think the macports version is needed anymore, and you can probably spend the time needed for maintaining it better on some other project.  I can handle an occasional outdated bug report, but if this becomes a common occurrence I will ask you to either update the macports version or to deprecate it.

**Comment by otfried** [comment]
Pinch zoom and two-finger pan works in 7.2.8.
